### PR TITLE
odc-geo migration

### PR DESCRIPTION
**Pull request by Ariana-B**
### Reason for this pull request

Geometry utilities have already been extracted from core into a separate package, `odc-geo`, however the code has not been replaced with calls to the `odc-geo` library. 

### Proposed changes

- Replace all `datacube.utils.geometry` imports with the corresponding `odc.geo` imports
- Remove geometry tests
- Mark datacube geometry utilities as deprecated but still maintain backwards compatibility for now
- Remove NetCDF Dataset wrapper
- Mark xarray `.geobox` property as deprecated

 - [ ] Closes #xxxx
 - [x] Tests added / passed
 - [x] Fully documented, including `docs/about/whats_new.rst` for all changes

----
:books: Documentation preview :books:: https://datacube-core--1424.org.readthedocs.build/en/1424/

### CODE REVIEW

**Comment by robbibt** [comment]
Hey @Ariana-B, really looking forward to this - will be fantastic to have `odc-geo` functionality available natively in datacube. A question from a user perspective: one of the most useful things about `odc-geo` for my work is being able to access `odc-geo` functionality via the `.odc` extension directly from an xarray dataset, which at the moment is achieved by importing `odc.geo.xr`: https://odc-geo.readthedocs.io/en/latest/intro-xr.html

Will it be possible to access that `.odc` extension on data loaded from the datacube with the changes you're making here too? For example, I'd love to be able to do something like this (without having to manually import `odc.geo.xr`):
```
import datacube

dc = datacube.Datacube()

ds = dc.load(product=..., x=..., y=..., time=...)

# Access odc-geo Geobox for loaded data
ds.odc.geobox
```

File: datacube/api/core.py [hunk 1]
```diff
@@ -1034,6 +1041,8 @@ def _make_dask_array(chunked_srcs,
     empties = {}  # type Dict[Tuple[int,int], str]
 
     def _mk_empty(shape: Tuple[int, ...]) -> str:
+        if type(shape) is not tuple:
```
Comment (SpacemanPaul) [inline datacube/api/core.py]: If this method accepts something other than Tuple you should update the typehint to reflect.

Presumably something like `Union[Tuple[int, ...], OtherType]`

File: datacube/drivers/netcdf/_safestrings.py
Comment (SpacemanPaul) [inline datacube/drivers/netcdf/_safestrings.py:1]: Note to other reviewers:  Ariana and I looked into it and the latest versions of the NetCDF library accept unicode strings, so this module is no longer required.

File: datacube/virtual/impl.py [hunk 2]
```diff
@@ -837,6 +837,7 @@ def fetch(self, grouped: VirtualDatasetBox, **load_settings: Dict[str, Any]) ->
 def reproject_band(band, geobox, resampling, dims, dask_chunks=None):
     """ Reproject a single measurement to the geobox. """
     if not hasattr(band.data, 'dask') or dask_chunks is None:
+        print(band.data.ndim)
```
Comment (SpacemanPaul) [inline datacube/virtual/impl.py]: Is this a leftover debugging trace?

File: docker/constraints.in [hunk 3]
```diff
@@ -38,6 +38,7 @@ sphinx_rtd_theme
 sqlalchemy>=2.0
 toolz
 xarray>=0.9
+odc-geo
```
Comment (SpacemanPaul) [inline docker/constraints.in:41]: You'll need to regenerate constraints.txt as well.

**Review by SpacemanPaul — CHANGES_REQUESTED** [review]
Overall, looking great.   See various minor inline comments, also you said "Mark datacube geometry utilities as deprecated but still maintain backwards compatibility for now" but that doesn't seem to have been done.

**Review by SpacemanPaul — APPROVED** [review]
Looks great, thanks Ariana.

As discussed, we'll give Kirill a chance to review before merging.

**Comment by robbibt** [comment]
Very excited about this PR! So at the moment (`datacube < 1.9`), the main way we access geo information about data loaded from the datacube is via the `.geobox` property that's automatically added to `xarray` data after you import `datacube`, e.g.:
```
ds.geobox
```
With `odc.geo.xr` imported, this geo information is automatically created via the new and more feature-rich `.odc.` extension instead, e.g.:
```
ds.odc.geobox
```

Is the plan that data loaded from datacube 1.9 will contain just the new automatically created `.odc.geobox` Geobox extension? Or will datacube 1.9 still add the older `.geobox` property to loaded data? 

My feeling is that we don't need both - we should rely on the `.odc.geobox` one created by `odc.geo` instead as it's much more powerful, and it will potentially be confusing to users to have the same geobox acessible from different places (although it probably depends on how much backwards compatibility matters to us for the 1.9 release, as removing `.geobox` is likely to break user code).

**Comment by Kirill888** [comment]
This [document](https://odc-geo.readthedocs.io/en/latest/migration.html) captures the main differences of `odc-geo` vs `datacube.geometry`. 

> via the .geobox property that's automatically added to xarray data after you import datacube

One could expose `.geobox` directly on an xarray for backwards compatibility using the following function in `odc-geo`

https://github.com/opendatacube/odc-geo/blob/b47103463f5beab27c48819744794e72c405bccc/odc/geo/_xr_interop.py#L762-L767

I think there is no issue having `import odc.geo.xr` present somewhere under `import datacube` code-path. It's not loaded by default in `odc-geo` imports because `xarray` is an optional and expensive dependency of `odc-geo`. This iss not the case with `datacube`.

Probably best to keep `.geobox` accessor for now, but it should generated deprecation warning on (first?) access. Attaching methods like this directly to xarray is now considered sub-optimal as there is now extension mechanisms present in xarray library.

File: datacube/utils/geometry/__init__.py [hunk 4]
```diff
@@ -44,7 +50,7 @@
     mid_longitude,
 )
 
-from .tools import (
+from .tools import (  # noqa
```
Comment (Kirill888) [inline datacube/utils/geometry/__init__.py:53]: Adding these to `__all__` below would remove the need for `# noqa`

**Review by Kirill888 — APPROVED** [review]
Thank you @Ariana-B  for going through this important but tedious change. I think this is good to merge. I would probably make sure that `xx.geobox` still works before making a release, it should issue deprecation warning stating that `xx.odc.geobox` is the new way to access this information.

File: datacube/utils/xarray_geoextensions.py [hunk 5]
```diff
@@ -7,23 +7,25 @@
...
 `.geobox`, `.affine` and `.extent` respectively.
 
 """
 import warnings
 import xarray
-from datacube.utils import geometry, spatial_dims
+from datacube.utils import spatial_dims
 from datacube.utils.math import affine_from_axis
+from odc.geo import CRS, CRSError, wh_
+from odc.geo.geobox import GeoBox
```
Comment (Kirill888) [inline datacube/utils/xarray_geoextensions.py]: One more thing that I have missed in the previous review. This file should just delegate geobox extraction to `.odc.geo.xr` with a warning, this extraction logic is in `odc-geo`.

**Comment by robbibt** [comment]
Thanks @Kirill888, happy with us just marking the current `.geobox` as deprecated for now.

On a similar note, should these also be marked as deprecated? They all seem like things we can access more reliably via `.odc.`
https://github.com/opendatacube/datacube-core/pull/1424/files/2eb5ef6c81b46939126208c2a344ce74d9f08e44#diff-32d508f00b999e6cf6606a0676f5f5262674086e8cef094545e42eb776630555L182-L187
```
xarray.Dataset.geobox = property(_xarray_geobox)    # type: ignore
xarray.Dataset.affine = property(_xarray_affine)    # type: ignore
xarray.Dataset.extent = property(_xarray_extent)    # type: ignore
xarray.DataArray.geobox = property(_xarray_geobox)  # type: ignore
xarray.DataArray.affine = property(_xarray_affine)  # type: ignore
xarray.DataArray.extent = property(_xarray_extent)  # type: ignore
```

File: datacube/utils/xarray_geoextensions.py (hunk 5 above)
Comment (Ariana-B) [reply datacube/utils/xarray_geoextensions.py]: I've implemented the change; let me know if it aligns with what you had in mind.

**Pull request merged**